Implement a `markersize` kwarg for scatter

chahak13 opened the pull request:
## PR Summary
Issue #1101 talks about having a kwarg called `markersize` (TBD) for the scatter function that can take sizes on a linear scale since it is currently on a quadratic scale with the kwarg `s`. The current implementation is a simple example of how this can be done directly on the collections level.

## PR Checklist

<!-- Please mark any checkboxes that do not apply to this PR as [N/A]. -->

**Documentation and Tests**
- [ ] Has pytest style unit tests (and `pytest` passes)
- [ ] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [ ] New plotting related features are documented with examples.

**Release Notes**
- [ ] New features are marked with a `.. versionadded::` directive in the docstring and documented in `doc/users/next_whats_new/`
- [ ] API changes are marked with a `.. versionchanged::` directive in the docstring and documented in `doc/api/next_api_changes/`
- [ ] Release notes conform with instructions in  `next_whats_new/README.rst` or `next_api_changes/README.rst`

greglucas commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -4588,13 +4588,17 @@ def scatter(self, x, y, s=None, c=None, marker=None, cmap=None, norm=None,
                 "or float array-like with the same size as x and y")
 
+        if markersize is not None:
+            s = np.ma.ravel(markersize)
```
what about squaring this, so it just cancels out the sqrt later? Then you wouldn't have to carry around the scaling.
```suggestion
            s = np.ma.ravel(markersize) ** 2
```

Do we always want to say that markersize takes precedence? Error/warn if both given?

Another option would be to have a `markerscaling` keyword argument that could be a string that controls whether to scale linearly or with the sqrt of the size.

chahak13 replied:
We'll for sure error out if both are given. I don't think it makes sense to give precedence to either. I just haven't implemented that because I just wanted to see if this idea works.

`markerscaling` is a good name. I like it.

chahak13 replied:
Wait, I think I misunderstood your `markerscaling` comment. Were you suggesting that we keep `s` as it is and just add `markerscaling` to control linear or quadratic?

greglucas replied:
I see your comment on the linked issue now about not wanting to square the size due to the inconsistency between set/get_sizes then, that makes sense to me, so probably ignore my comment about squaring above.

Yes, my comment meant to make `markerscaling='linear'` take a string, which would then just change how `s=[5, 6, 7]` is interpreted, rather than having a user choose whether to put their data in `markersize` or `size`.  I guess another option would be to do something like `markerscaling=1`, and then `np.sqrt(np.power(s, markerscaling))` in the collection... Which is similar to what you have here already.

chahak13 replied:
One of the reasons for the issue and this PR is to have a more readable name for marker sizes too as `s` has no meaning by itself. I guess if `size` were allowed, this would not be that much of a problem but it's not. On the other hand, `plot` allows `markersize` to manage the sizes of the markers when not a line so having something similar here would also give more consistency in terms of usage. (I don't know how many times I've had to refer the documentation because I tried markersize with scatter first).

jklymak commented:
I think this misses the semantic difference between `s` in `scatter` and `markersize` in `plot`.  In `scatter`, `s` is meant to be an array in `scatter`, whereas `markersize` cannot be an array in `plot`.  I think the changes proposed here make the difference even more confusing, and we should be very deliberate about what we do here.  

First, I don't think you should snake the scaling through to the low-level object.  get_sizes should just return what it's always returned, with an explanation in the docs.  

Second, I think you should consider that `s=array` is how to specify arrays for the markersizes, and `markersize=float` is a shortcut for `plot`-like scaling if `s` is not provided.  eg, `markersize` should _not_ accept an array.  If `s` and `markersize` are both specified, it should probably error out.

chahak13 commented:
> I don't think you should snake the scaling through to the low-level object. get_sizes should just return what it's always returned, with an explanation in the docs.

My only concern with this is if the user has provided the sizes in the linear scale, then it makes sense to return the sizes that were "set" instead of the modified ones. But this goes hand in hand with the discussion of what we want `markersize` to be doing. 

There are two things that the original issue/discussion covered: 1. Having `markersize` (or better) as a more user-friendly kwarg and 2. To provide a way to have linearly scaling sizes. For 2, it makes more sense to me to be able to add an array for `markersize` than to not because that is taking away functionality that can be provided without much effort. If the consensus is to not allow arrays, then that should not be difficult.

I agree with not allowing both `s` and `markersize` at the same time. It does that already, that's the first check.

jklymak commented:
My tendency is to be pretty conservative when it comes to multiple ways to do the same thing.  I'd be very concerned about the Container returning different values for size based on how the axes helper created it.  Yes, you have a new flag on the Container, but downstream libraries don't know that, so if anyone is depending on the markersize being in one units or another, they are not going to know to check the markerscale.

chahak13 commented:
I understand that, but wouldn't having the default behaviour such that it doesn't change anything that already exists resolve that issue? And if they are instead using `markersize` then they would know already what will be returned based on the docs?

chahak13 commented:
Meeting 9/3: Think instead about making `markersize` an alias for `s` and have a scaling factor as the kwarg to control all this. Also look into how to make this scaling factor consistent with all the getter/setters.

jklymak commented:
@chahak13 does this still need discussion or do you have enough guidance?

chahak13 commented:
@jklymak sorry, I should've put a comment in earlier but I've changed `markersize` to be an alias for `s` and added a `markerscale` variable as we discussed last time in the meeting.

chahak13 commented:
Sorry for the label, was looking for something else!

timhoffm commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -4568,6 +4568,12 @@ def scatter(self, x, y, s=None, c=None, marker=None, cmap=None, norm=None,
             or ``nan``). If ``True`` the points are drawn with the *bad*
             colormap color (see `.Colormap.set_bad`).
 
```
I would not call this just scale, because it could be interpreted as "linear scaling factor" - and we do this e.g. at `mutation_scale` in https://matplotlib.org/devdocs/api/_as_gen/matplotlib.patches.FancyArrowPatch.html#matplotlib.patches.FancyArrowPatch. Suggestion: `markerscaling` or maybe just `scaling` - I have some sympathy to push this down to collectionWithSizes` later. A `scaling` parameter would be meaningful there so that we have consistent naming and don't need to translate the markerscaling kwarg. Also, I believe scatter(..., scaling=...) is clear enough.

Also, 1 and 2 are not very telling as values. How about linear/quadratic or linear/square or length/area?
